### PR TITLE
[KLC-2358] expose redundancy metrics via /node/status and Prometheus

### PR DESCRIPTION
**Pull request by fbsobreira**
## Summary
Expose `klv_redundancy_level`, `klv_redundancy_is_active`, and (for backups) `klv_redundancy_is_main_active` via both `/node/status` JSON and the `/node/metrics` Prometheus surface. Replaces a prior design where a single role-dependent metric was misread as "main producer down" for healthy main nodes.

## Problem
Three related gaps existed:

1. **Prometheus blind to redundancy state.** Previous wiring injected `klv_redundancy_level` directly into the `/node/status` JSON map in `network/api/node/routes.go`, bypassing `statusMetrics`. The Prometheus serializer reads from that same store, so the metric never appeared in `/node/metrics`.
2. **Inactive node silently misreported.** In the klever-connector path, `node → JSON → connector → presenter` decodes JSON numbers as `float64`. The connector cast them via `uint64(v)`, and `uint64(float64(-1.0))` is implementation-defined per the Go spec — on `gc/amd64` it yields `0`. A node started with `--redundancy-level=-1` was displayed as a main producer.
3. **Role-confused metric.** An initial fix added `klv_redundancy_is_main_active` mirroring the `slotsOfInactivity < 5*level` formula. The formula is backup-perspective ("is my upstream alive?") but the metric was emitted for every role. For a level=0 main the formula always returned 0, so healthy main producers reported `is_main_active=0` and triggered false alerts.

## Solution
**Role-aware metrics, role-agnostic interpretation.** Two booleans, each answering one question with the same meaning for every role:

| Metric | Role applicability | What it answers |
|---|---|---|
| `klv_redundancy_level` | Every node | What's my configured role? (-1 inactive / 0 main / N>0 backup-rank-N) |
| `klv_redundancy_is_active` | Every node | Am I currently producing for my slots? (1 yes / 0 standby or inactive) |
| `klv_redundancy_is_main_active` | **Backups only** (`level > 0`) | Is my upstream chain alive? (1 alive / 0 silent — this backup is taking over) |
| `klv_redundancy_slots_inactive` | Every node | Slots elapsed since last upstream activity (unchanged from develop) |

Computation is role-aware inside `buildNodeMetricsPollingFunc` and reads `(level, slotsOfInactivity, isMainMachineActive)` atomically via a new `NodeRedundancyHandler.Snapshot()` under a single RLock, so the published triple cannot disagree across separate getter calls.

Operator queries become single-liners with no role filter:

```promql
# Who's producing right now?
klv_redundancy_is_active == 1

# Is exactly one node producing per shard?
count by (shard) (klv_redundancy_is_active == 1)

# Failover happening (a backup is producing instead of the main)
klv_redundancy_is_active == 1 and klv_redundancy_level > 0

# Backup upstream-health monitoring (only meaningful for level > 0)
klv_redundancy_is_main_active == 0
```

For the connector, signed metrics dispatch through `SetInt64Value` via a `signedMetricKeys` allowlist; the float→int casts are now guarded against `NaN`, `±Inf`, and out-of-range values (using `>=` against `MaxInt64` / `MaxUint64` because float64 cannot represent those bounds exactly).

## Key Changes
- **New:**
  - `klv_redundancy_is_active` metric — every node.
  - `klv_redundancy_is_main_active` metric — backups only (emitted iff `level > 0`).
  - `NodeRedundancyHandler.Snapshot() (int64, uint64, bool)` — atomic read of level/slots/upstreamAlive under one RLock.
  - `core.redundancy.GetInternalRedundancyLevel()` — RLocked accessor for parity with sibling getters.
  - `signedMetricKeys` allowlist in `cmd/connector/provider/metricsProvider.go`.
  - `getFromCacheAsInt64` helper in `statusHandler/presenter/common.go` with `MaxInt64`-bounded uint64 fallback.
- **Updated:**
  - `cmd/node/metrics/metrics.go` — polling closure now writes all redundancy metrics through the shared status-handler store, primed synchronously before the recurring loop so request #1 sees real values.
  - `statusHandler/presenter/instanceInfoGetters.go` — typed int64/uint64 lookups (`GetRedundancyLevel`, `GetRedundancyIsActive`). The presenter returns `"N/A"` when the metric is absent so the TUI hides the redundancy row before the first poll.
  - `statusHandler/view/termuic/termuiRenders/widgetsRender.go` — main and inactive nodes show only their role; backups append `(standby)` or `(taking over)` derived from `is_active`.
  - `core/metrics.go` — godoc on both new constants documents the role-agnostic value semantics and (for `is_main_active`) the backup-only emission contract.
- **Removed:**
  - `network/api/node.RedundancyLevelKey` — orphan constant; the same string value is available as `core.MetricRedundancyLevel`.
  - The inline `details[RedundancyLevelKey] = facade.GetRedundancy()` injection (superseded by the polling path).
  - `GetRedundancy()` method on `network/api/node.FacadeHandler` interface — no remaining consumers in the package.

## Testing
- [x] All existing tests pass for touched packages.
- [x] `go vet ./...` clean on changed files (pre-existing `kvm/wasmer2/*` `unsafe.Pointer` warnings are unrelated).
- [x] Race detector clean on the new tests.
- [x] New tests:
  - `TestSetPresenterValue_ConnectorPath_PreservesNegativeRedundancyLevel` — full JSON → presenter pipeline; no shortcuts that bypass `float64`.
  - `TestSetPresenterValue_RejectsOutOfRangeFloats` — NaN, ±Inf, above MaxInt64 / below MinInt64 / above MaxUint64 / negative-on-unsigned all rejected with `ErrTypeAssertionFailed`.
  - `TestBuildNodeMetricsPollingFunc_AppearsInStatusMetricsAndPrometheus` — pins both metrics appearing in JSON and Prometheus.
  - `TestBuildNodeMetricsPollingFunc_PublishesRedundancyMetrics` — table-driven (main / backup-standby / backup-taking-over / inactive) covering all 4 role × upstream-state combinations, including the assertion that `is_main_active` is **absent** for non-backups.
  - `TestInitMetrics_DoesNotPreInitRedundancyIsActive` — pins the deliberate "hide-until-polled" contract.
  - `TestPresenterStatusHandler_GetRedundancyLevel` / `GetRedundancyIsActive` — typed-getter coverage including legacy uint64 fallback, MaxInt64-overflow rejection, and out-of-range `{0,1}` → N/A.
  - `TestComputeRedundancyStr` — five role × is_active states for the TUI renderer, including hidden-row pre-first-poll and edge cases (level=-100, level=100).

### Manual smoke test
```bash
go build -o klever-node ./cmd/node

# main producer
./klever-node ... --redundancy-level=0
# backup rank 2
./klever-node ... --redundancy-level=2
# permanently inactive
./klever-node ... --redundancy-level=-1
```

Expected values per role:

| Role | level | is_active | is_main_active |
|---|---|---|---|
| `--redundancy-level=0` main, healthy | `0` | `1` | _absent_ |
| `--redundancy-level=2` backup, upstream alive | `2` | `0` | `1` |
| `--redundancy-level=2` backup, upstream silent (taken over) | `2` | `1` | `0` |
| `--redundancy-level=-1` inactive | `-1` | `0` | _absent_ |

```bash
curl -s http://127.0.0.1:8080/node/status | jq '.data.metrics | {
  klv_redundancy_level,
  klv_redundancy_is_active,
  klv_redundancy_is_main_active,
  klv_redundancy_slots_inactive
}'

curl -s http://127.0.0.1:8080/node/metrics | grep klv_redundancy_
```

## Configuration Changes
None. `--redundancy-level` flag is unchanged.

## Breaking Changes
- **Removed exported Go symbols:**
  - `network/api/node.RedundancyLevelKey` — use `core.MetricRedundancyLevel` (same string value).
  - `GetRedundancy() int64` method on the local `network/api/node.FacadeHandler` interface — the method is still defined on `common/facade.NodeHandler` for other callers.
- **New `Snapshot() (int64, uint64, bool)` method on `consensus.NodeRedundancyHandler`** — third-party implementers (none known) must add it. Both in-tree implementer and mock stub are updated.
- **New JSON / Prometheus surface:** `klv_redundancy_is_active` (every node) and `klv_redundancy_is_main_active` (backups only). The metric `klv_redundancy_is_main_active` was a defined constant on `develop` but never written, so no released binary emitted it; consumers depending on the old never-written semantic do not exist.
- **The on-the-wire format of `klv_redundancy_level`** is unchanged (still JSON int64 / Prometheus numeric).

## Operator notes
- `/node/status` and `/node/metrics` expose redundancy topology (level + role flags). Operators with public-facing nodes should either restrict the endpoints to a private network / management VLAN, or set `Secured=true` on the routes — otherwise an attacker can map the backup chain and mount a targeted DoS.
- Alerts on `klv_redundancy_is_active == 0` will fire on permanently-inactive observers too. Filter on `klv_redundancy_level >= 0` if you only care about validators, or use the simpler `klv_redundancy_is_active == 1 and klv_redundancy_level > 0` to surface failover events specifically.

## Related Issues
Fixes KLC-2358

## Affected blockchain-critical components
- Consensus: consensus.NodeRedundancyHandler gained Snapshot(); core/redundancy enforces RLock for GetInternalRedundancyLevel and adds Snapshot() to atomically return (redundancyLevel, slotsOfInactivity, isMainMachineActive). Consensus mocks updated to implement Snapshot().
- Networking / API surface: redundancy metrics are moved onto the StatusMetrics store and exposed via /node/status (JSON) and /node/metrics (Prometheus). network/api/node no longer injects redundancy directly into the route; exported RedundancyLevelKey constant and FacadeHandler.GetRedundancy() were removed.
- Connector (metrics ingestion): cmd/connector/provider now uses a signedMetricKeys allowlist and guarded float→int conversions, routing signed metrics via SetInt64Value and rejecting NaN/±Inf/out-of-range floats.
- Node metrics subsystem & presenter/view: cmd/node/metrics now primes and uses a single polling closure that reads redundancy via Snapshot(), computes role-aware metrics (MetricRedundancyLevel, MetricRedundancyIsActive for all nodes, MetricRedundancyIsMainActive only for backups), and publishes them atomically. Presenter helpers and view interfaces changed to consume the new metrics (getFromCacheAsInt64, GetRedundancyIsActive, Presenter interface updated).
- State/heartbeat mocks: node/heartbeat mock RedundancyHandlerStub implements Snapshot().

## Impact on node stability and data integrity
- Fixes a data-integrity bug in the connector where JSON-decoded float64 → unsigned casts could silently convert negative redundancy levels to zero (mislabeling inactive nodes as mains). Signed metrics now use int64 with explicit bounds checks; all float inputs reject NaN/Inf/out-of-range values to avoid implementation-defined casts.
- Atomic Snapshot() prevents torn reads of related redundancy state (level, inactivity slots, upstream alive) that could otherwise produce inconsistent metric combinations and lead to incorrect takeover/alerting behavior.
- Metrics publication is role-aware and atomic within the polling closure, reducing race-induced metric churn and false alerts (e.g., previously emitting backup-perspective main-active formulas for all roles).

## Concurrency and error-handling changes
- Added RLock/RUnlock protection and a Snapshot() reader to ensure consistent multi-field reads under a single lock.
- Connector path now explicitly validates numeric inputs and returns errors for unsupported types or out-of-range float values; presenter getters now return "N/A" for missing/invalid metrics to hide rows until first poll.
- Presenter cache access gains getFromCacheAsInt64 with legacy uint64 fallback guarded against overflow to avoid wraparound.

## Cross-cutting concerns and testing
- Presenter/view changes adjust TUI formatting: redundancy row hidden until MetricRedundancyIsActive is first published; computeRedundancyStr updated to format role-aware strings.
- Tests and stubs updated/added across connector, metrics polling, presenter getters, and render formatting; mock stubs adapted to provide Snapshot() fallback for backwards-compatible test behavior.
- Race-detector/vet-conscious fixes: tighter float bound checks and mock changes to satisfy updated interfaces.

## Breaking changes / migration notes
- Removed exported constant: network/api/node.RedundancyLevelKey (klv_redundancy_level).
- Removed FacadeHandler.GetRedundancy() method.
- Added Snapshot() to consensus.NodeRedundancyHandler interface — implementations must provide atomic snapshot support.
- New JSON/Prometheus metrics: klv_redundancy_is_active and klv_redundancy_is_main_active (semantics documented in core/metrics.go).

## Operational effect
- Behavior change: redundancy metrics are now accurate, role-aware, and appear in both /node/status and Prometheus; TUI hides redundancy until metrics are polled. These changes improve correctness of takeover coordination and alerting and remove a source of false-positive main/backup state reporting.

[Review Change Stack](https://app.coderabbit.ai/change-stack/klever-io/klever-go/pull/52?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: statusHandler/presenter/instanceInfoGetters.go [hunk 1]
```diff
@@ -30,21 +29,39 @@ func (psh *PresenterStatusHandler) GetPublicKeyBlockSign() string {
...
+
+	valUint64, ok := val.(uint64)
+	if !ok {
+		return metricNotAvailable
+	}
+
+	if valUint64 == 1 {
+		return "true"
+	}
+	return "false"
```
[the remark this answers is not in the document]
Comment (fbsobreira) [reply statusHandler/presenter/instanceInfoGetters.go]: Good catch — accepted. The writer in `cmd/node/metrics.buildNodeMetricsPollingFunc` only ever publishes 0 or 1, but the getter is the safety boundary, and silently coercing an unknown upstream value to `"false"` would misreport an unknown state as a failover signal. Switched the branch to an explicit `switch` with `1 → "true"`, `0 → "false"`, default `metricNotAvailable` so the TUI hides the row instead. Added a regression test `uint64 outside {0,1} returns N/A instead of false`.

File: statusHandler/presenter/common.go [hunk 2]
```diff
@@ -23,6 +23,31 @@ func (psh *PresenterStatusHandler) getFromCacheAsUint64(metric string) uint64 {
...
+	if !ok {
+		return 0
+	}
+
+	switch v := val.(type) {
+	case int64:
+		return v
+	case uint64:
+		return int64(v) // #nosec G115: only lossless for non-negative writers
+	default:
```
[the remark this answers is not in the document]
Comment (fbsobreira) [reply statusHandler/presenter/common.go:43]: Accepted. The redundancy-level domain stays far below `math.MaxInt64` in practice, but the bound is a 1-line guard that removes the foot-gun for any future signed metric with a wider range. Added `if v > math.MaxInt64 { return 0 }` on the uint64 fallback, updated the godoc, and added a regression test (`uint64 above MaxInt64 returns 0 instead of wrapping negative`).

File: cmd/node/metrics/metrics.go [hunk 3]
```diff
@@ -95,6 +95,10 @@ func InitMetrics(
 	appStatusHandler.SetUInt64Value(core.MetricNodeUptimeSeconds, initUint)
 	appStatusHandler.SetUInt64Value(core.MetricNodeStartTimestamp, initUint)
 	appStatusHandler.SetUInt64Value(core.MetricRedundancySlotsInactive, initUint)
+	appStatusHandler.SetInt64Value(core.MetricRedundancyLevel, 0)
+	// MetricRedundancyIsMainActive is intentionally NOT pre-initialised: the TUI
+	// presenter treats "key absent" as N/A and hides the redundancy row until the
+	// first polling tick writes the real boolean value (within pollingInterval).
```
[the remark this answers is not in the document]
Comment (fbsobreira) [reply cmd/node/metrics/metrics.go]: You're right — the comment described pre-prime behaviour. `StartNodeMetricsPolling` now calls `pollingFunc(ash)` synchronously before `Poll()`, so the metric is published as soon as polling starts rather than after a `pollingInterval` delay. Reworded the comment to match.

File: cmd/node/metrics/matics_test.go [hunk 4]
```diff
@@ -545,6 +599,124 @@ func TestStartNodeMetricsPolling_Valid(t *testing.T) {
...
+	assert.Contains(t, promOut, core.MetricRedundancyLevel+"{")
+	assert.Contains(t, promOut, core.MetricRedundancyIsMainActive+"{")
+	// Sanity: the value must be emitted, not just the label. Match "} <value>" rather
+	// than the trailing suffix so the assert survives any future label additions.
+	for _, line := range strings.Split(promOut, "\n") {
+		if strings.HasPrefix(line, core.MetricRedundancyIsMainActive+"{") {
+			assert.Contains(t, line, "} 0", "expected uint64 0, got: %q", line)
+		}
+		if strings.HasPrefix(line, core.MetricRedundancyLevel+"{") {
+			assert.Contains(t, line, "} 2", "expected int64 2, got: %q", line)
```
[the remark this answers is not in the document]
Comment (fbsobreira) [reply cmd/node/metrics/matics_test.go]: Respectfully declining: testify's `assert.Contains` does interpolate printf-style `msgAndArgs`. From [`assertions.go` (`messageFromMsgAndArgs`)](https://github.com/stretchr/testify/blob/master/assert/assertions.go):

```go
if len(msgAndArgs) > 1 {
    return fmt.Sprintf(msgAndArgs[0].(string), msgAndArgs[1:]...)
}
```

So `assert.Contains(t, line, "} 0", "expected uint64 0, got: %q", line)` formats correctly on failure. The `Containsf` variant is a syntactic-clarity helper, not a correctness requirement. Leaving the assert as-is.

File: cmd/connector/provider/metricsProvider.go [hunk 5]
```diff
@@ -111,8 +126,23 @@ func (smp *StatusMetricsProvider) applyMetricsToPresenter(metricsMap map[string]
...
+		if _, signed := signedMetricKeys[key]; signed {
+			if v < math.MinInt64 || v > math.MaxInt64 {
+				return ErrTypeAssertionFailed
+			}
+			smp.presenter.SetInt64Value(key, int64(v))
+			return nil
+		}
+		if v < 0 || v > math.MaxUint64 {
+			return ErrTypeAssertionFailed
+		}
```
[the remark this answers is not in the document]
Comment (fbsobreira) [reply cmd/connector/provider/metricsProvider.go]: Accepted as a 2-char change. You're right that `float64(math.MaxInt64)` and `float64(math.MaxUint64)` both round up (to 2^63 and 2^64), so the previous `> math.MaxInt64` / `> math.MaxUint64` allowed the rounded-up boundary value through and would have hit the same implementation-defined cast we're trying to avoid. Switched both bounds to `>=` and noted the float-precision rationale in the inline comment. Practical reachability remains zero (JSON integers above 2^53 are already lossy), but the fix costs nothing.

File: cmd/connector/provider/metricsProvider.go [hunk 6]
```diff
@@ -111,8 +126,25 @@ func (smp *StatusMetricsProvider) applyMetricsToPresenter(metricsMap map[string]
...
+			if v < math.MinInt64 || v >= math.MaxInt64 {
+				return ErrTypeAssertionFailed
+			}
+			smp.presenter.SetInt64Value(key, int64(v))
+			return nil
+		}
+		if v < 0 || v >= math.MaxUint64 {
+			return ErrTypeAssertionFailed
+		}
 		smp.presenter.SetUInt64Value(key, uint64(v))
```
[the remark this answers is not in the document]
Comment (fbsobreira) [reply cmd/connector/provider/metricsProvider.go:148]: Good catch. Added `math.Trunc(v) != v` between the NaN/Inf check and the bounds check so fractional floats are rejected explicitly, and extended `TestSetPresenterValue_RejectsOutOfRangeFloats` with `1.9` (signed, MetricRedundancyLevel) and `2.5` (unsigned, MetricNumTxInBlock) cases to lock the invariant.

Producer-side typed setters make a decimal value unreachable today, but enforcing the contract at the connector boundary is cheaper than debugging it from a misreported metric later. Verified with `go test -race ./cmd/connector/provider/`.

**Review by Beroni — APPROVED** [review]

**Review by RomuloSiebra — APPROVED** [review]
